src/Cardano/AddressV3.cpp

```diff
@@ -0,0 +1,234 @@
+        throw std::invalid_argument("Wrong spending key size");
+    }
+    if (groupKey.size() != 32) {
+        throw std::invalid_argument("Wrong group key size");
+    }
+    auto addr = AddressV3();
+    addr.discrimination = discrimination_in;
+    addr.kind = Kind_Group;
+    addr.key1 = spendingKey;
+    addr.groupKey = groupKey;
+    return addr;
+}
+
+AddressV3 AddressV3::createAccount(Discrimination discrimination_in, const Data& accountKey) {
+    if (accountKey.size() != 32) {
+        throw std::invalid_argument("Wrong spending key size");
+    }
+    auto addr = AddressV3();
+    addr.discrimination = discrimination_in;
+    addr.kind = Kind_Account;
+    addr.key1 = accountKey;
+    return addr;
+}
+
+AddressV3::AddressV3(const std::string& addr) : legacyAddressV2(nullptr) {
+    if (parseAndCheckV3(addr, discrimination, kind, key1, groupKey)) {
+        // values stored
+        return;
+    }
+    // try legacy
+    // throw on error
+    legacyAddressV2 = new TW::Cardano::AddressV2(addr);
+}
+
+AddressV3::AddressV3(const PublicKey& publicKey) : legacyAddressV2(nullptr) {
+    // input is extended pubkey, 64-byte
+    if (publicKey.type != TWPublicKeyTypeED25519Extended) {
+        throw std::invalid_argument("Invalid public key type");
+    }
+    discrimination = Discrim_Test;
+    kind = Kind_Group;
+    key1 = Data(32);
+    groupKey = Data(32);
+    std::copy(publicKey.bytes.begin(), publicKey.bytes.begin() + 32, key1.begin());
+    std::copy(publicKey.bytes.begin() + 32, publicKey.bytes.begin() + 64, groupKey.begin());
+}
+
+AddressV3::AddressV3(const Data& data) : legacyAddressV2(nullptr) {
+    // min 4 bytes, 2 prefix + 2 len
+    if (data.size() < 4) { throw std::invalid_argument("Address data too short"); }
+    assert(data.size() >= 4);
+    int index = 0;
+    discrimination = (Discrimination)data[index++];
+    kind = (Kind)data[index++];
+    // key1:
+    byte len1 = data[index++];
+    if (data.size() < 4 + len1) { throw std::invalid_argument("Address data too short"); }
+    assert(data.size() >= 4 + len1);
+    key1 = Data(len1);
+    std::copy(data.begin() + index, data.begin() + index + len1, key1.begin());
+    index += len1;
+    // groupKey:
+    byte len2 = data[index++];
+    if (data.size() < 4 + len1 + len2) { throw std::invalid_argument("Address data too short"); }
+    assert(data.size() >= 4 + len1 + len2);
+    groupKey = Data(len2);
+    std::copy(data.begin() + index, data.begin() + index + len2, groupKey.begin());
+}
+
+AddressV3::AddressV3(const AddressV3& other) :
+    discrimination(other.discrimination),
+    kind(other.kind),
+    key1(other.key1),
+    groupKey(other.groupKey),
+    legacyAddressV2(other.legacyAddressV2 == nullptr ? nullptr : new AddressV2(*other.legacyAddressV2))
+{}
+
+string AddressV3::string() const {
+    std::string hrp;
+    switch (kind) {
+        case Kind_Single:
+        case Kind_Group:
+        case Kind_Account:
+            hrp = "addr"; break;
```

Comment:
let's add `hrp` to coins.json?

Reply:
Yes, done.  The switch is still needed, though.